tests/test_dfa.py

```diff
@@ -1453,3 +1453,67 @@ def test_dfa_repr(self):
         self.assertEqual(
             repr(dfa),
             "DFA(states={'q0'}, input_symbols={'a'}, transitions={'q0': {'a': 'q0'}}, initial_state='q0', final_states={'q0'}, allow_partial=False)")  # noqa: E501
+
+    def test_contains_subsequence(self):
+        dfa = DFA.contains_subsequence('', {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.contains_subsequence('0', {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.contains_subsequence('01', {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.contains_subsequence('010', {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.contains_subsequence('dcba', {'a', 'b', 'c', 'd'})
+        self.assertFalse(dfa.isfinite())
+
+    def test_of_length(self):
+        dfa = DFA.of_length(0, float('inf'), {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.of_length(4, float('inf'), {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.of_length(0, 8, {'a', 'b'})
+        self.assertTrue(dfa.isfinite())
+
+        dfa = DFA.of_length(4, 8, {'0', '1', '2'})
+        self.assertTrue(dfa.isfinite())
+
+    def test_nth_from_end(self):
+        with self.assertRaises(ValueError):
+            dfa = DFA.nth_from_end('0', 0, {'0', '1'})
+
+        with self.assertRaises(ValueError):
+            dfa = DFA.nth_from_end('2', 1, {'0', '1'})
+
+        dfa = DFA.nth_from_end('0', 1, {'0'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.nth_from_end('0', 1, {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.nth_from_end('0', 2, {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.nth_from_end('0', 3, {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+        dfa = DFA.nth_from_end('1', 4, {'0', '1'})
+        self.assertFalse(dfa.isfinite())
+
+    def test_empty_language(self):
```

Comment:
Higher up in this file, there's a test that uses DeMorgan's laws along with the empty and universal languages. It would be good to use the functions you wrote in those tests. I think there may also be other tests that use the empty language (and include redundant generator code for those DFAs).

Reply:
Yeah I'm thinking a lot of tests can be updated since many of the DFA we test with are either "contains this pattern" or "avoids this pattern" where the patterns are either substrings or subsequences.

Reply:
I've updated some of the old tests. I didn't want to change anything to do with equality since in those it's best to have some manual constructions.